arch/arm/core/aarch64/smp.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+/**
+ * @file
+ * @brief codes required for AArch64 multicore and Zephyr smp support
+ */
+
+#include <cache.h>
+#include <device.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <soc.h>
+#include <init.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <drivers/interrupt_controller/gic.h>
+#include <drivers/psci.h>
+#include <sys/arch_interface.h>
+
+volatile struct {
+	arch_cpustart_t fn;
+	void *arg;
+} arm64_cpu_init[CONFIG_MP_NUM_CPUS];
+
+volatile char *arm64_cpu_sp;
+
+volatile uint32_t arm64_cpu_up_flag;
+
+/*
+ * _curr_cpu is used to record the struct of _cpu_t of each cpu.
+ * for efficient usage in assembly
+ */
+volatile _cpu_t *_curr_cpu[CONFIG_MP_NUM_CPUS];
+
+extern void __start(void);
+/* Called from Zephyr initialization */
+void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+		    arch_cpustart_t fn, void *arg)
+{
+	const struct device *psci;
+
+	_curr_cpu[cpu_num] = &(_kernel.cpus[cpu_num]);
+	arm64_cpu_init[cpu_num].fn = fn;
+	arm64_cpu_init[cpu_num].arg = arg;
+
+	arm64_cpu_sp = Z_THREAD_STACK_BUFFER(stack) + sz;
+	arch_dcache_range((void *)&arm64_cpu_sp, sizeof(char *),
+			  K_CACHE_WB_INVD);
+
+	arm64_cpu_up_flag = cpu_num;
+	arch_dcache_range((void *)&arm64_cpu_up_flag, sizeof(uint32_t),
+			  K_CACHE_WB_INVD);
+
+	psci = device_get_binding("PSCI");
+	/* TODO: get mpidr from device tree, using cpu_num */
+	if (psci_cpu_on(psci, cpu_num, (uint64_t)&__start))
```

Comment:
Is `psci_cpu_on` supposed to be a memory barrier? Is `psci_cpu_on` supposed to ensure memory coherency if e.g. the other core comes up with its mmu disabled?

Reply:
Ok. In that case is `psci_cpu_on` supposed to ensure memory ordering here?

Reply:
`arch_dcache_range`  takes care to ensure `arm64_cpu_init[cpu_num] ` is populated *before*  starting the `cpu_num`.  Do you see any other ordering requirement here.